Bluetooth: controller: implement ISO flush timeout

kruithofa opened the pull request:
Bluetooth: controller: implement ISO flush timeout
This commit implements the flush timeout for ISO as per BT
spec 5.4 Vol 6. Part B section 4.5.13.5

fixes #42901

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: controller: implement ISO flush timeout

This commit implements the flush timeout for ISO as per BT
spec 5.4 Vol 6. Part B section 4.5.13.5

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral_iso.c:
```diff
@@ -647,13 +684,10 @@ static void isr_rx(void *param)
-
-		/* Start transmitting new burst */
-		cis_lll->tx.bn_curr = 1U;
+		cis_lll->tx.bn_curr = 1; /* FIXME shouldn't this be done in the prepare_cb? */
```
revert to LHS (i see no change done than remove existing comment).

You could add additional comment that, as a peripheral bn_curr is reset after Radio Rx (instead of as in central where it transmits first in an event)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -572,6 +572,20 @@ void ll_cis_create(uint16_t cis_handle, uint16_t acl_handle)
+		cis->lll.tx.ft_cntr_se = cis->lll.tx.ft *
+			DIV_ROUND_UP(cis->lll.nse, cis->lll.tx.bn);
+	} else {
+		cis->lll.tx.ft_cntr_se = 0;
```
Would assigning to `1U` make any difference. I prefer `1U` if there is no mention that it shall be 0 when bn is 0.

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral_iso.c:
```diff
@@ -268,6 +268,22 @@ uint8_t ull_peripheral_iso_acquire(struct ll_conn *acl,
+		cis->lll.rx.ft_cntr_se = 0;
+	}
+
+
```
We use the following for peripheral, if it's any help. based on BT Core 5.2, Vol 6, Part B, section 4.5.13.5:
E = floor(cisPayloadNumber / BN)
U = NSE - floor(NSE / BN) * (BN - 1 - (cisPayloadNumber mod BN))

E:  (E + FT - 1) = CIS event where flush point occurs
U:  Flush point sub event
FE: Flush point event
```
flush_event = (payload_number / rxtx->bn) + rxtx->ft - 1;
flush_subevents = nse - (nse / rxtx->bn) * (rxtx->bn - 1 - (payload_number % rxtx->bn));
```
                       
`flush_event`  is the event count at which the payload is flushed
`flush_subevents` is the number of subevents into `flush_event` where flush point occurs

This passes qualification.

kruithofa replied:
Thanks, I will  make sure that we follow the same

Thalley commented:
@Chris-StJohn-Bose have a look at this PR :)

cvinayak committed:
Bluetooth: Controller: FT decremented by NSE remaining for Central ISO

Decrement flush timeout subevent countdown by remaining
subevents when flushing Tx for Central ISO role.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented:
@kruithofa @cvinayak since both this PR and https://github.com/zephyrproject-rtos/zephyr/pull/59222 are closed, do we have another PR that fixes the issue, or what is going on? :D

cvinayak commented:
> @kruithofa @cvinayak since both this PR and #59222 are closed, do we have another PR that fixes the issue, or what is going on? :D

https://github.com/zephyrproject-rtos/zephyr/pull/59409 is the new draft PR that is passing CI, and handles FT >= 1 but for BN=1. Work is still in progress to add FT >= 1 with BN >= 1 implementation. Other corner cases like events being aborted, and skipped etc needs implementation, under these cases the ISO data get NACK-ed etc... for now in the PR
